### PR TITLE
Proposal - Schema for Batch Messages

### LINKED ISSUE #120: [Batched Messages] Spike: Define Schema for Batched Messages
# Things to Figure Out

As a user, I want to publish a batch of messages, so that others can read those messages.

## Acceptance Criteria
- [ ] How do others know how to read the data I share?
    - Schemas
    - Data format
    - How do two independent actors go from MRC Schema with no messages each arrive at the same Parquet file given the same messages?
- [ ] How do others know the data I shared is from me or a delegate?
    - Validation
    - Delegation (MRC Batches don't support delegation, that is up the protocol to define aka the DSNP Id)
- [ ] How I let other people know that I created a batch file and how they can get and read it?
    - Who: Announcer MSA Id
    - Where: IFPS URI
    - What: Schema Id
    - How: File Format
- [ ] As a user who wants to read messages, I want to be able to call an rpc to get a paginated list of batch file metadata (aka ipfs uris) for a particular data structure (aka a given Schema Id)
- [ ] (Not this story) As the network, we want to limit spam even if that spam is off-chain. To that end we want to charge capacity for off-chain messages
    - Current Solution: Make announcers provide the file size along with the other metadata for the batch. Charge extra capacity based on this file size. Consumers of that batch file are only expected to download the file size number of bytes and anything more than that should be disregard such that lying about file size is useless as others will merely receive bad files.

### Properties of Batches

We think this all leads to Parquet files (with compression and bloom filters)

- Low bandwidth
- Small storage
- Easy to parse
- Searchable/filterable on header data download
- Scalable to large number of entries

# OLD

Consumer of network should be provided the schema defining a batch announcement

# Acceptance Criteria
- [ ] Document and present options for handling batch reference posted to the chain

# Description
See [BatchAnnouncementParams](https://github.com/LibertyDSNP/mrc/blob/main/designdocs/batched_messages.md#types)
For more details on a batch announcement schema types

Things to consider/Need Details:
- How do we structure the data for the batch reference posted to the chain?
  - e.g. Should batch announcement type schema be registered with schema pallet and if so define the schema type as well as publish schema_id
  - Should the structure be completely fixed
  - How does this schema get registered, if it's stored in Schema storage in the schema pallet -- Genesis config?

Remember: Schemas have an choice at creation:
1. The avro data for each message is stored on-chain
2. The data for each message is in a batch

### PR DESCRIPTION
**Pull request by dopatraman**
# Goal
The purpose of this PR is to propose a schema for batch messages.

The design is still in flux.

Closes #120

### CODE REVIEW

File: designdocs/batched_messages.md [hunk 1]
```diff
@@ -62,6 +62,86 @@ See the [implementation of paging in the messages pallet](https://github.com/Lib
     * `next_index`: `u32` starting index of next results in `next_block`
     * `results`: `Vec<BatchAnnouncement>`
 
+### Message Schema
+
+Distributing data on a decentralized network has a few fundamental problems that
+are worth discussing here. Primarily, those problems are:
+- What shape is this data in?
+- How do I authenticate this data?
+- How do these actors access this information?
```
Comment (johnmela) [inline designdocs/batched_messages.md:71]: I found this confusing until reading further down.

Maybe:
`How can actors access this information?` or
`How can users access this information?`

File: designdocs/batched_messages.md [hunk 2]
```diff
@@ -62,6 +62,86 @@ See the [implementation of paging in the messages pallet](https://github.com/Lib
...
+
+Less primary but very much salient are a couple of secondary concerns:
+- How do I make sure this data occupies no more space / time than necessary?
+- How do I make sure this data can be easily indexed and searched?
+
+In the days of yore, it was easy to answer these questions. A centralized
+service would declare an output shape and make it available to the world via an
+API they would maintain and authorize. In our world, it is not so simple. Each
+actor on a DSNP-implementing system must be able to imbue their announcements
+with trust without exerting control over the entire network. Announcements can
```
Comment (johnmela) [inline designdocs/batched_messages.md:81]: Will `Announcent` need to be modified to match the nomenclature defined in the Frequency Chain diagram or is this related to DSNP spec?

File: designdocs/batched_messages.md [hunk 3]
```diff
@@ -62,6 +62,86 @@ See the [implementation of paging in the messages pallet](https://github.com/Lib
...
+would not serve its users needs.
+
+Let's try addressing the above concerns one-by-one.
+
+#### Message Shape
+
+The batch file message schema could look like the following:
+
+```json
+// Batch Message
```
Comment (johnmela) [inline designdocs/batched_messages.md:93]: Same with this, should `Batch Message` match whatever nomenclature is decide in @wilwade 's  diagram? 

There are other terms below that will likely not be decided upon until next week, so rather than comment each one I'll just leave this one as a reminder that we should probably have whatever terminology is decided upon as canon, to be matched in code and documentation.

File: designdocs/logical_batching.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+   - Off-chain / On-chain Payload
+
+*Protocols* would be responsible for indicating not only the shape of the
+underlying payload, but also its format and location. This is important, because
+a protocol's format property would indicate whether a given payload contains a
+batch or not.
+
+## FormatType
+
+Payloads can either be of format type `PARQUET` or `AVRO`. These file formats
```
Comment (shannonwells) [inline designdocs/logical_batching.md:28]: Are these going to be enums?

File: designdocs/logical_batching.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+deriving the idea of a batch logically, as a product of fields present on other
+entities in our system.
+
+## Background
+
+We are currently experimenting with a new system architecture that has the
+following entities:
+
+1. Schema
+   - Id
```
Comment (shannonwells) [inline designdocs/logical_batching.md:14]: All these fields need types.  Also, are we going to have a "pay more" type of field? We were originally going to use the file size but @wilwade suggested it be something more general. Don't know what it would be called.

File: designdocs/logical_batching.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+   - Format
+   - FormatType
+   - PayloadLocation
+2. Message
+   - Off-chain / On-chain Payload
+
+*Protocols* would be responsible for indicating not only the shape of the
+underlying payload, but also its format and location. This is important, because
```
Comment (shannonwells) [inline designdocs/logical_batching.md:22]: We went back to using Schema instead of Protocol, based on what I heard at doc day yesterday.

File: designdocs/logical_batching.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+| Avro   | IPFS (Off-chain) | Private Graph (?) |
+| Parquet| IPFS (Off-chain) | Announcement      |
+```
+
+(More needed here).
+
+## Message Payloads
+
+Message payloads can be either on-chain or off-chain. If they are off-chain,
+there is a good chance that they will be located on IPFS.
```
Comment (shannonwells) [inline designdocs/logical_batching.md:50]: Off-chain file URIs should be IPFS CIDv2 as shown here: https://github.com/LibertyDSNP/mrc/blob/main/designdocs/batched_messages.md#types
I'm not aware of any indecision about that; having a URI be content addressable is a requirement.

File: designdocs/logical_batching.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+We can circumvent defining a batch explicitly if we leverage the file format
+included in the protocol. Since both Avro and Parquet file types declare whether
+or not they contain single or plural values, the format itself could indicate
+whether or not the message is a batch.
+
+## Rationale
+
+There are some upsides to deriving batching logically from existing structures.
+One is costsavings. Not having a batch structure means we don't need to worry
```
Comment (shannonwells) [inline designdocs/logical_batching.md:62]:
```suggestion
One is cost savings. Not having a batch structure means we don't need to worry
```

File: designdocs/logical_batching.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
+# Summary
+
+In other documents, we explore ways to structure an explicit batch of messages.
+That is, in terms of concrete structure. This document has a different aim --
+deriving the idea of a batch logically, as a product of fields present on other
+entities in our system.
```
Comment (shannonwells) [inline designdocs/logical_batching.md:6]: So are we registering a Batch Message Schema using these fields?  Would that be the story to write from this document?

File: designdocs/logical_batching.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+## FormatType
+
+Payloads can either be of format type `PARQUET` or `AVRO`. These file formats
+indicate specifically whether their payloads are singular or arrays of objects
+(More info needed here).
+
+## PayloadLocation
+
+The combination of format and location constrain possible payload types:
```
Comment (shannonwells) [inline designdocs/logical_batching.md:34]: This could use a little more elaboration.

**Review by shannonwells — CHANGES_REQUESTED** [review]
Just be sure to use the design doc template - not all sections will apply but most will.  One should be able to write stories for coding to a design doc once it's finished.

More specific comments inline.

File: designdocs/logical_batching.md (hunk 5 above)
Comment (wilwade) [reply designdocs/logical_batching.md:14]: Re: pay more / files size

I think we should stick with `payload_size` (perhaps `payload_size_bytes`?) for now, although for clarity that goes under message (line 18/19) not Schema (which is where some form of pay more setup might go)

File: designdocs/logical_batching.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+```txt
+| Format  | Location        | Payload           |
+-------------------------------------------------
+| Avro   | On-chain         | Public Graph      |
+| Parquet| On-chain         | ??                |
+| Avro   | IPFS (Off-chain) | Private Graph (?) |
+| Parquet| IPFS (Off-chain) | Announcement      |
+```
```
Comment (wilwade) [inline designdocs/logical_batching.md:43]: The payload column should just be Examples. Perhaps something like:
```suggestion
```txt
| Format  | Location        | Examples           |
-------------------------------------------------
| Avro   | On-chain         | DSNP Graph Change      |
| Parquet| On-chain         | Unknown                |
| Avro   | IPFS (Off-chain) | Larger Avro structures |
| Parquet| IPFS (Off-chain) | DSNP Broadcast or Reply Announcements      |
```
```

File: designdocs/logical_batching.md (hunk 9 above)
Comment (wilwade) [reply designdocs/logical_batching.md:6]: I'm working on writing those stories. Here is the list I have so far to do:

- Story to add Format Type to the Schema (pointing to an enum with just Avro currently)
- Story to document Avro Format Type requirements
- Story to add Payload Location (on-chain only enum) to the Schema
  - Dependent-story to add the off-chain to the enum and support it in the message_pallet
- Story to add Parquet to the Format Type enum and document Parquet Format Type requirements

Comment (wilwade) [reply designdocs/logical_batching.md:6]: https://github.com/LibertyDSNP/mrc/milestone/11

File: designdocs/logical_batching.md (hunk 10 above)
Comment (dopatraman) [reply designdocs/logical_batching.md:34]: Agreed, I'm not sure what else to say here though. I'm open to suggestions.

File: designdocs/logical_batching.md (hunk 9 above)
Comment (shannonwells) [reply designdocs/logical_batching.md:6]: This design doc needs to specify what fields are in a Batch message, including types.  If it's not settled then there's no Design Doc to write. If it's settled, it should be documented so that the stories can refer back to the Design Doc for implementation.

The Design Doc template is also part of the process, it's intended to help write the document.  Not all sections may need to be there (glossary or table of contents are two obvious examples)

**Review by shannonwells — CHANGES_REQUESTED** [review]

File: designdocs/logical_batching.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+## Background
+
+We are currently experimenting with a new system architecture that has the
+following entities:
+
+1. Schema
+   - id: u16
+   - format: Format
+   - format_type: FormatType
```
Comment (johnmela) [inline designdocs/logical_batching.md:16]: Reminder to modify this to match whatever the decision is for what this will end up being called. I think we said `Model` and `ModelType` yesterday.

File: designdocs/batched_messages.md [hunk 13]
```diff
@@ -10,92 +10,148 @@
...
+This document specifes how messages can be inferred from both schema format type
+and payload location.
 
-This document does not describe the types of DSNP messages that would be in batch announcements. Batch announcements must reference a schema through its ID number. A schema may or may not describe a DSNP message.
+This document also specifies how schemas will constrain the shape of off chain
+messages.
 
-This document does not discuss validation of batch file format located at the URI in the announcement, since
-the file format cannot be verified on-chain. For details about batch files themselves, see [DSNP Spec: Batch Publications](https://spec.dsnp.org/DSNP/BatchPublications).
+This document does not describe the types of DSNP messages that will be
```
Comment (wilwade) [inline designdocs/batched_messages.md:43]:
```suggestion
This document does not describe the types of DSNP Announcements that will be
```

File: designdocs/batched_messages.md [hunk 14]
```diff
@@ -10,92 +10,148 @@
...
-The file hash is not included separately. Since the `batch_uri` uses CIDv1 specification, the file hash is already included.
+* `ModelType` - supported serialization formats for message payloads files. Currently only [Parquet](https://parquet.apache.org/docs/) and
+  [Avro](https://avro.apache.org/docs/current/) are supported.
+* `PayloadLocation` - The location of the payload. Can be either `OnChain` or `IPFS`.
 
-* `BatchAnnouncement`: implements `BatchAnnouncementParams`, returned by `get_batches`
+### Traits
+* `Model` - TBD. A set of functions for accessing message payload information.
+  * Derives `Encode`, `Decode`, `MaxEncodedLen`
+  * `max_length`: `SchemaMaxBytesBoundedVecLimit`
```
Comment (wilwade) [inline designdocs/batched_messages.md:65]: I'm not sure what this trait is for.

File: designdocs/batched_messages.md [hunk 15]
```diff
@@ -10,92 +10,148 @@
...
-    * `from_index`: `u32`, starting message index
-    * `page_size`: `usize`, retrieve `page_size` messages at a time, up to configured `T::BatchPageSizeMax`. If 0, return `T::BatchPageSizeMax` results
-
-* `BatchAnnouncementResult`
-    * `has_next`: `uint32`, current page number
-    * `next_block`: `<T::BlockNumber>` starting block number of next page of results
-    * `next_index`: `u32` starting index of next results in `next_block`
-    * `results`: `Vec<BatchAnnouncement>`
+### Types
+* `SchemaAnnouncement<T:Config, M: Model>`: generic
```
Comment (wilwade) [inline designdocs/batched_messages.md:68]:
```suggestion
* `Schema<T:Config, M: Model>`: generic
```

File: designdocs/batched_messages.md [hunk 16]
```diff
@@ -10,92 +10,148 @@
...
+    * `model_type`: `ModelType` See enum section above.
+    * `model`: `M` Defines the shape of the message payload.
+    * `payload_location`: `PayloadLocation` See enum section above.
+
+* `MessageAnnouncement<T:Config>`: generic
```
Comment (wilwade) [inline designdocs/batched_messages.md:73]:
```suggestion
* `Message<T:Config>`: generic
```

File: designdocs/batched_messages.md [hunk 17]
```diff
@@ -10,92 +10,148 @@
...
+    * `schema_id`: `u16`
+    * `source`: `MsaId` Source of the message.
+    * `provider`: `MsaId` Public key of a capacity-providing account
+    * `onchain_payload`: `Vec<u8>?`
+    * `offchain_payload`: `Vec<u8>?`
```
Comment (wilwade) [inline designdocs/batched_messages.md:78]: Agreed that this is still a bit TBD.

In the end we have this data:

- OnChain
  - `source`: `MsaId` Source of the message.
  - `payload`: `Vec<u8>` (I think)
- IPFS
  - `payload_cid`: `?` CIDv2 of the IPFS file
  - `payload_byte_length`: `u32 ?` The size of the IPFS file in bytes

How we store that data in the message I think is an outstanding question. I think there are several options, but not sure what way we should go.

File: designdocs/batched_messages.md [hunk 18]
```diff
@@ -10,92 +10,148 @@
...
 ### Extrinsics
-#### announce_batch(origin, batch_announcement_params)
-Creates and posts a new batch announcement message on chain, using the batch message Schema Id. This Schema Id must already be registered and will need to be fetched by the extrinsic.  The transaction fee is determined in part by the file size.
+#### register_schema(origin, schema_params)
+Creates and posts a new schema on chain. The transaction fee is determined in part by the model size.
 
 * **Parameters**
   * origin:  required for all extrinsics, the caller/sender.
-  * `batch_announcement_params`: `BatchAnnouncementParams`, the parameters to use in the batch announcement.
+  * `schema_params`: `SchemaAnnouncement`, the parameters to use in the batch announcement.
```
Comment (wilwade) [inline designdocs/batched_messages.md:86]:
```suggestion
  * `schema_params`: `Schema`, the parameters to use in the batch announcement.
```

File: designdocs/batched_messages.md [hunk 19]
```diff
@@ -10,92 +10,148 @@
...
 
-* **Event**:  `Event::<T>::BatchAnnounced(schema_id, msa_id, file_size, batch_uri)`
+* **Event**:  `Event::<T>::SchemaAnnounced(schema_id, model, model_type, payload_location)`
```
Comment (wilwade) [inline designdocs/batched_messages.md:88]: I don't think we have any new events other than what is already in the Message design doc.

SchemaCreated is already a thing, and when the messages design doc covers the events when a new message is created.

File: designdocs/batched_messages.md [hunk 20]
```diff
@@ -10,92 +10,148 @@
...
 * **Restrictions**:
-  * origin's `msa_id` must have capacity to post the transaction during the current epoch
+  * TBD
 
 ### Custom RPCs
 
-#### get_batches(options)
-Retrieves paged batch announcements that have not been garbage-collected which meet `options` criteria.  The `msa_id` does not need to be active.
+#### get_schema(schema_id)
+Retrieves a `SchemaAnnouncement`.
```
Comment (wilwade) [inline designdocs/batched_messages.md:95]:
```suggestion
Retrieves a `Schema`.
```

File: designdocs/batched_messages.md [hunk 21]
```diff
@@ -10,92 +10,148 @@
...
 
 * **Parameters**
-  * `options`: `BatchAnnouncementOptions` containing batch announcement criteria
+  * `schema_id`: `u16` a schema identifier
 
 * **Returns**
-  * `None()` if no messages meet the criteria.
-  * `Some(Vec<BatchAnnouncement>)`, in descending block-transaction order
+  * `None()` if no schemas meet the criteria.
+  * `Some(SchemaAnnouncement)`
```
Comment (wilwade) [inline designdocs/batched_messages.md:102]:
```suggestion
  * `Some(Schema)`
```

File: designdocs/batched_messages.md [hunk 22]
```diff
@@ -10,92 +10,148 @@
...
+[Avro docs](https://avro.apache.org/docs/current/spec.html#schemas)). These files
+could be stored either on chain or off chain. If they are on chain, it would
+make sense for the file to be small (lower cost). However, they could be large
+and stored off chain.
+
+See below to see how the combination of format and location indicate possible
+payload types:
+
+```txt
+| Model Type | Location         | Payload                               |
```
Comment (wilwade) [inline designdocs/batched_messages.md:124]:
```suggestion
| Model Type | Location         | Example Use Case                        |
```

File: designdocs/batched_messages.md [hunk 23]
```diff
@@ -10,92 +10,148 @@
 * [Glossary](#glossary)
 
 ## Context and Scope
-This design document describes batched messages and APIs for sending and retrieving them.
+This design document describes message schemas. It also will describe
+batchability as a logical construct derived from schemas.
+
+We will also be describing APIs for sending and retrieving messages.
```
Comment (wilwade) [inline designdocs/batched_messages.md:16]:
```suggestion
We will also be updating the APIs for creating schemas.
```

**Review by wilwade — CHANGES_REQUESTED** [review]
Most of the change requests and alignment on naming.

The one big question is how to handle the different fields needed by `Message` based on the schema properties.

File: designdocs/batched_messages.md (hunk 14 above)
Comment (wilwade) [reply designdocs/batched_messages.md:65]: Ah I see as `max_length` might be different for different Model Types right?

Comment (dopatraman) [reply designdocs/batched_messages.md:65]: yes, that and payload

File: designdocs/batched_messages.md [hunk 24]
```diff
@@ -10,98 +10,198 @@
...
+* `ModelType` - supported serialization formats for message payloads files. Currently only [Parquet](https://parquet.apache.org/docs/) and
+  [Avro](https://avro.apache.org/docs/current/) are supported.
+* `PayloadLocation` - The location of the payload. Can be either `OnChain` or `OffChain`.
+  * `OnChain`
+  * `OffChain`
+* `Payload`
+  * `OnChain`
+    * `source`: `MsaId`
+    * `payload`: `Vec<u8>`
+  * `IPFS`
```
Comment (aramikm) [inline designdocs/batched_messages.md:64]: any reason why we don't need `source` for IPFS?

File: designdocs/batched_messages.md [hunk 25]
```diff
@@ -10,98 +10,198 @@
...
 
 We discussed whether to allow URLs such as HTTP/HTTPS or other URLs and instead opted for content-addressable URIs (CIDv1) which can be resolved by some other service.  This allows us to put the file hash directly into a URI.  It reduces message storage because we don't have to include both a URL and a file hash. A file hash is necessary as a check against file tampering.
 
 We revisited the idea of whether it really is necessary to include a file size. We will be charging a premium for larger files, however, there will be per-byte discount for larger files in order to create an incentive for posting batches while reducing the incentive for announcers to allow spam. Although the processing and downloading time for enormous files also serves as a disincentive for spam, we feel it would not be sufficient.
 
-Despite the fact that announcers can lie abut the file size, the file_size parameter also serves as an on-chain declaration that not only allows consumers of batches to quickly discover if a batch announcer was honest, but the file requestor can know in advance when to stop requesting data.
+Despite the fact that announcers can lie abut the file size, the file_size
```
Comment (aramikm) [inline designdocs/batched_messages.md:166]:
```suggestion
Despite the fact that announcers can lie about the file size, the file_size
```

File: designdocs/batched_messages.md [hunk 26]
```diff
@@ -10,98 +10,198 @@
...
+}
+
+trait IPFSPayload {
+  // FYI: the following is just for demonstrative purposes
+  fn payload_cid() -> CIDv2
+  fn payload_length() -> u32
+}
+
+enum Payload {
+  OnChain(Box<dyn OnChainPayload>),
```
Comment (aramikm) [inline designdocs/batched_messages.md:194]: nit: I don't think using `dyn` is recommended. I think we can rewrite this in a way to remove `Box` and `dyn`

**Review by aramikm — COMMENTED** [review]
Overall looks good, added some comments and suggestions

File: designdocs/batched_messages.md [hunk 27]
```diff
@@ -10,98 +10,198 @@
 * [Glossary](#glossary)
 
 ## Context and Scope
-This design document describes batched messages and APIs for sending and retrieving them.
+This design document describes message schemas. It also will describe
+batchability as a logical construct derived from schemas.
+
+We will also be updating the APIs for creating schemas`.
```
Comment (wilwade) [inline designdocs/batched_messages.md:16]:
```suggestion
We will also be updating the APIs for creating schemas.
```

File: designdocs/batched_messages.md [hunk 28]
```diff
@@ -10,98 +10,198 @@
...
+network congestion, announcers collate messages into batches of the same type of
+message and announce the batch location on-chain, instead of announcing each
+individual message.
+
+However, this idea does not go far enough. Batching allows us to support posting
+massive amounts of data, but it would still be expensive to post data of this
+size on chain.
+
+We can leverage off chain storage to make posting large message collections cheap, but we need to
+find a way to constrain the message types posted offchain so that message
```
Comment (wilwade) [inline designdocs/batched_messages.md:29]:
```suggestion
find a way to constrain the message types posted off chain so that message
```

File: designdocs/batched_messages.md [hunk 29]
```diff
@@ -10,98 +10,198 @@
...
-* `BatchSizeMinBytes` - the minimum possible size of a valid batch file with 1 row
-
 ### Enums
-* `BatchFormat` - a list of supported formats for batch files. Currently only Parquet file format is supported, however,
-other formats are being considered.
+* `ModelType` - supported serialization formats for message payloads files. Currently only [Parquet](https://parquet.apache.org/docs/) and
+  [Avro](https://avro.apache.org/docs/current/) are supported.
+* `PayloadLocation` - The location of the payload. Can be either `OnChain` or `OffChain`.
+  * `OnChain`
+  * `OffChain`
```
Comment (wilwade) [inline designdocs/batched_messages.md:59]: Swapping to match below.
```suggestion
  * `IPFS`
```

File: designdocs/batched_messages.md [hunk 30]
```diff
@@ -10,98 +10,198 @@
...
-### Constants
-* `BatchPageSizeMax` - the maximum number of batch announcements that will be returned by the `get_batches` query
-* `BatchSizeMinBytes` - the minimum possible size of a valid batch file with 1 row
-
 ### Enums
-* `BatchFormat` - a list of supported formats for batch files. Currently only Parquet file format is supported, however,
-other formats are being considered.
+* `ModelType` - supported serialization formats for message payloads files. Currently only [Parquet](https://parquet.apache.org/docs/) and
+  [Avro](https://avro.apache.org/docs/current/) are supported.
+* `PayloadLocation` - The location of the payload. Can be either `OnChain` or `OffChain`.
```
Comment (wilwade) [inline designdocs/batched_messages.md:57]: Swapping to match below. (Other off chain locations may be supported in the future)
```suggestion
* `PayloadLocation` - The location of the payload. Can be either `OnChain` or `IPFS`.
```

File: designdocs/batched_messages.md [hunk 31]
```diff
@@ -10,98 +10,198 @@
...
+  * `OnChain`
+  * `OffChain`
+* `Payload`
+  * `OnChain`
+    * `source`: `MsaId`
+    * `payload`: `Vec<u8>`
+  * `IPFS`
+    * `payload_cid`: `TBD`
+    * `payload_byte_length`: `TBD`
```
Comment (wilwade) [inline designdocs/batched_messages.md:66]: I think we can set this to `u64` safely

File: designdocs/batched_messages.md [hunk 32]
```diff
@@ -10,98 +10,198 @@
...
+  [Avro](https://avro.apache.org/docs/current/) are supported.
+* `PayloadLocation` - The location of the payload. Can be either `OnChain` or `OffChain`.
+  * `OnChain`
+  * `OffChain`
+* `Payload`
+  * `OnChain`
+    * `source`: `MsaId`
+    * `payload`: `Vec<u8>`
+  * `IPFS`
+    * `payload_cid`: `TBD`
```
Comment (wilwade) [inline designdocs/batched_messages.md:65]: 1. Should likely be clear in that it is a v1 cid
2. We can use `Vec<u8>` here I think. That's what https://docs.rs/cid/0.8.5/cid/struct.CidGeneric.html#method.to_bytes returns.
```suggestion
    * `payload_cidv1`: `Vec<u8>`
```

File: designdocs/batched_messages.md (hunk 24 above)
Comment (wilwade) [reply designdocs/batched_messages.md:64]: On-chain data can have this source specified and validated on-chain, but off-chain data, it is not as simple.

That said I wonder if we should shift "on-chain delegation" (aka source enabled/disabled) to a boolean flag (via enum for readability perhaps) on the schema. @dopatraman thoughts?

For those schemas not using batching like DSNP, they might have all that data coming from a delegated source.

It feels complex, but perhaps we actually reduce the complexity overall. First, around payload because then source is "always" there (just sometimes required to be empty by the schema flag for it) and then the payload is simpler. Second, we don't have an AND. The payload location doesn't ALSO define if source is a thing, which isn't really a part of payload location.

File: designdocs/batched_messages.md [hunk 33]
```diff
@@ -10,98 +10,198 @@
...
-    * `to_block`: `<T::BlockNumber>`, retrieve messages ending at the given block number (inclusive)
-    * `from_index`: `u32`, starting message index
-    * `page_size`: `usize`, retrieve `page_size` messages at a time, up to configured `T::BatchPageSizeMax`. If 0, return `T::BatchPageSizeMax` results
-
-* `BatchAnnouncementResult`
-    * `has_next`: `uint32`, current page number
-    * `next_block`: `<T::BlockNumber>` starting block number of next page of results
-    * `next_index`: `u32` starting index of next results in `next_block`
-    * `results`: `Vec<BatchAnnouncement>`
+(See alternatives section for another way to structure payloads)
```
Comment (wilwade) [inline designdocs/batched_messages.md:85]: I don't actually understand this way of structuring it. The alt section I do, but I don't understand why 
    source, payload, payload_cid, and payload_byte_length are under enum.

File: designdocs/batched_messages.md [hunk 34]
```diff
@@ -10,98 +10,198 @@
...
   * origin:  required for all extrinsics, the caller/sender.
-  * `batch_announcement_params`: `BatchAnnouncementParams`, the parameters to use in the batch announcement.
+  * `schema_params`: `Schema`, the parameters to use in the batch announcement.
 
-* **Event**:  `Event::<T>::BatchAnnounced(schema_id, msa_id, file_size, batch_uri)`
 * **Restrictions**:
-  * origin's `msa_id` must have capacity to post the transaction during the current epoch
+  * TBD
 
 ### Custom RPCs
```
Comment (wilwade) [inline designdocs/batched_messages.md:98]: I think we need to understand how `message.get_messages_by_schema` rpc will change.

File: designdocs/batched_messages.md [hunk 35]
```diff
@@ -10,98 +10,198 @@
...
-    * `page_size`: `usize`, retrieve `page_size` messages at a time, up to configured `T::BatchPageSizeMax`. If 0, return `T::BatchPageSizeMax` results
-
-* `BatchAnnouncementResult`
-    * `has_next`: `uint32`, current page number
-    * `next_block`: `<T::BlockNumber>` starting block number of next page of results
-    * `next_index`: `u32` starting index of next results in `next_block`
-    * `results`: `Vec<BatchAnnouncement>`
+(See alternatives section for another way to structure payloads)
 
 ### Extrinsics
```
Comment (wilwade) [inline designdocs/batched_messages.md:87]: We need to add the updated signature for `messages.add` here as well.

File: designdocs/batched_messages.md [hunk 36]
```diff
@@ -10,98 +10,198 @@
...
+The document above describes payload location as an enumeration of on-chain and
+off-chain types. This allows us to encapsulate different information for both
+on-chain and IPFS payloads without overloading a single type with both sets of
+fields.
+
+However, if we discover that both on-chain and off-chain payloads are too
+complex to fit into a single enum, we could use traits to our advantage:
+
+```rust
+trait OnChainPayload {
```
Comment (wilwade) [inline designdocs/batched_messages.md:181]: I don't know that we ever need to read this data on-chain, just write (or delete). The `get_messages_by_schema` rpc call does need to be able to build the correct response type however. Hmm... Mostly just thinking out loud

**Review by wilwade — DISMISSED** [review]
Few typos I found while re-reading, a few more pieces to add around message wr, and some thoughts on the payload structuring

File: designdocs/batched_messages.md (hunk 24 above)
Comment (wilwade) [reply designdocs/batched_messages.md:64]: On additional conversation, If we shift that to a separate flag we can make that a separate story or design doc.

File: designdocs/batched_messages.md (hunk 34 above)
Comment (dopatraman) [reply designdocs/batched_messages.md:98]: if I understand correctly, it may not need to change? It looks like the only `Message` related struct is `MessageResponse`, which will keep the same structure it has now. Please let me know if there's something I'm missing.

Comment (wilwade) [reply designdocs/batched_messages.md:98]: Hmm... I think that is correct for now. We can improve that later.

File: designdocs/batched_messages.md [hunk 37]
```diff
@@ -10,98 +10,235 @@
...
+`payload_location`.
 
-### Benefits and Risks
-Please see [Batching Source Dependent Messages With Delegation](https://forums.projectliberty.io/t/04-batching-source-dependent-messages-with-delegation/216), for discussion about the benefits of announcing batch files on chain rather than all types of user-created messages.
+* **Parameters**
+  * `origin`: `Origin` A signed transaction origin from the provider
+  * `on_behalf_of`: `Option<MessageSourceId>` The msa id of delegate.
+  * `schema_id`: `u16` A schema identifier
+  * `payload`: `Payload` The message payload
+  * `payload_location`: `PayloadLocation` The payload location
```
Comment (wilwade) [inline designdocs/batched_messages.md:121]: I'm not sure why we need payload_location. That information should be derivable from `schema_id` right?

**Review by wilwade — APPROVED** [review]
One question on the suggested implementation, but I think we are good to merge. I expect those implementation details will be easier to iron out in the stories and code.

File: designdocs/batched_messages.md [hunk 38]
```diff
@@ -10,98 +10,235 @@
...
 We revisited the idea of whether it really is necessary to include a file size. We will be charging a premium for larger files, however, there will be per-byte discount for larger files in order to create an incentive for posting batches while reducing the incentive for announcers to allow spam. Although the processing and downloading time for enormous files also serves as a disincentive for spam, we feel it would not be sufficient.
 
-Despite the fact that announcers can lie abut the file size, the file_size parameter also serves as an on-chain declaration that not only allows consumers of batches to quickly discover if a batch announcer was honest, but the file requestor can know in advance when to stop requesting data.
+Despite the fact that announcers can lie about the file size, the file_size
+parameter also serves as an on-chain declaration that not only allows consumers
+of batches to quickly discover if a batch announcer was honest, but the file
+requestor can know in advance when to stop requesting data.
+
+#### Message Payloads
+The document above describes payload location as an enumeration of on-chain and
```
Comment (shannonwells) [inline designdocs/batched_messages.md:188]: I wonder if since these implementation details are so up in the air whether it's worth outlining them at all. Maybe we could just say that the batch/schema updates will affect the current message retrieval format and that the specifics are TBD by implementer(s).  It's a good suggestion but maybe belongs in a story for that instead (i.e. don't just delete the ideas)

File: designdocs/batched_messages.md [hunk 39]
```diff
@@ -10,98 +10,235 @@
...
+  [Avro](https://avro.apache.org/docs/current/) are supported.
+* `PayloadLocation` - The location of the payload. Can be either `OnChain` or `IPFS`.
+  * `OnChain`
+  * `IPFS`
+* `Payload`
+  * `OnChain`
+    * `source`: `MsaId`
+    * `payload`: `Vec<u8>`
+  * `IPFS`
+    * `payload_cidv1`: `Vec<u8>`
```
[the remark this answers is not in the document]
Comment (wilwade) [reply designdocs/batched_messages.md:65]: I pushed for cidv1. That way a single schema can only support one location. If you want to support a schema you know exactly how to get the data and you don't end up in a resolver mess and half supported uris

File: designdocs/batched_messages.md (hunk 37 above)
Comment (dopatraman) [reply designdocs/batched_messages.md:121]: good point
